### DISCUSSION TITLE
Example: How to set up a password hashing routine for an Entity / ActiveModel

### DISCUSSION THREAD

**Discussion by mohs8421** [SeaQL/sea-orm#678] [Q&A] [answered]
Hi,

I'm new to sea-orm, and just setting up a new system.
Obviously I need some kind of login and want to set it up nicely in the user entity.
Currently it looks like this, but I am unsure if this is the best way to go, and it seems to not to be utilizing the features of sea-orm properly:
```
impl ActiveModel {
    pub fn hash_password(password: &str) -> String {
        let argon2 = Argon2::default();
        argon2
            .hash_password(password.as_bytes(), &SaltString::generate(OsRng))
            .expect("Generating password hash failed.")
            .to_string()
    }

    pub fn set_password(&mut self, password: &str) {
        self.password = ActiveValue.Set(self::hash_password(password));
    }
}
```
So how would this look like if done right?

**Comment by billy1624** [comment, marked as answer]
Hey @mohs8421, welcome! You can use the `before_save` hook in `ActiveModelBehavior`. See https://www.sea-ql.org/SeaORM/docs/generate-entity/expanded-entity-structure#active-model-behavior

**Reply by mohs8421** [reply]
I tried it, but it does not execute the before_save action.
Here is the testing code
```
async fn add_user(db: &DatabaseConnection, email: String) {
    println!("Adding new user {}", email);

    let mut new_user = entity::user::ActiveModel {
        user_id: ActiveValue::NotSet,
        email: ActiveValue::Set(email),
        password: ActiveValue::NotSet,
    };
    let generated_password = new_user.generate_password();
    entity::user::Entity::insert(new_user)
        .exec(db)
        .await
        .expect("Inserting record failed.");
    println!(
        "New user has been added, password generated was '{}'",
        generated_password
    );
}

//...

impl ActiveModelBehavior for ActiveModel {
    fn before_save(mut self, _insert: bool) -> Result<Self, DbErr> {
        println!("This is the current password: {:?}", self.password);
        if let ActiveValue::Set(new_pw) = self.password {
            println!("password is getting hashed now!");
            self.password = ActiveValue::Set(hash_password(new_pw));
        }
        Ok(self)
    }
}
```
It does not print the debug information.

**Reply by mohs8421** [reply]
honestly I would prefer it, if the type of the password is not String but a PasswordHash.

**Reply by billy1624** [reply]
Try `new_user.insert(db)`

**Reply by mohs8421** [reply]
interesting loophole, isn't this a bug?

**Reply by billy1624** [reply]
`entity::user::Entity::insert` operate on the executor level which doesn't support event hook

**Reply by fcoury** [reply]
Wow, wasted maybe 5 hours of my life on this. Would be nice to have a note about this somewhere in the documentation.

Thanks @billy1624 for the answer.

**Reply by alvissraghnall** [reply]
> Hey @mohs8421, welcome! You can use the `before_save` hook in `ActiveModelBehavior`. See https://www.sea-ql.org/SeaORM/docs/generate-entity/expanded-entity-structure#active-model-behavior

Doesn't this run on updates as well? If yes, say you updated an entity, it would compute an entirely different hash of the already hashed password, no?
